### PR TITLE
Generalize DataOut::write_vtu for simplex mesh

### PR DESCRIPTION
**Pull request by peterrum**
... reusing functionalities from #10701.

### CODE REVIEW

File: source/base/data_out_base.cc [hunk 1]
```diff
@@ -828,6 +890,48 @@ namespace
       }
   }
 
+  template <int dim, int spacedim>
+  static void
+  compute_sizes(const std::vector<DataOutBase::Patch<dim, spacedim>> &patches,
+                const bool    write_higher_order_cells,
+                unsigned int &n_nodes,
+                unsigned int &n_cells,
+                unsigned int &n_points_an_n_cell)
```
Comment (drwells) [inline source/base/data_out_base.cc:899]: Should this be `n_points_and_n_cells`? I find the abbreviation hard to read.

Comment (drwells) [reply source/base/data_out_base.cc:899]: The name predates this PR but I would still like to fix it.

File: source/base/data_out_base.cc [hunk 2]
```diff
@@ -5028,33 +5129,11 @@ namespace DataOutBase
...
-          vtk_cell_id = {24, 1};
-        else if (patch.reference_cell_type == ReferenceCell::get_hypercube(dim))
-          vtk_cell_id = {vtk_cell_type[dim],
-                         Utilities::pow(patch.n_subdivisions, dim)};
-        else
-          {
-            Assert(false, ExcNotImplemented());
-          }
+        const auto vtk_cell_id =
+          extract_vtk_patch_info(patch, flags.write_higher_order_cells);
```
Comment (drwells) [inline source/base/data_out_base.cc:5133]: Much nicer!

**Review by drwells — CHANGES_REQUESTED** [review]
Lets rename the variable - otherwise :+1:

File: source/base/data_out_base.cc (hunk 1 above)
Comment (peterrum) [reply source/base/data_out_base.cc]: I fixed this typo!

**Review by drwells — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Pull request merged**